File: extensions/shared-youtube/library/src/main/java/app/morphe/extension/shared/patches/SanitizeSharingLinksPatch.java
package app.morphe.extension.shared.patches;

import app.morphe.extension.shared.privacy.LinkSanitizer;
import app.morphe.extension.shared.settings.SharedYouTubeSettings;

/**
 * YouTube and YouTube Music.
 */
@SuppressWarnings("unused")
public final class SanitizeSharingLinksPatch {

    private static final LinkSanitizer sanitizer = new LinkSanitizer(
            "si",
            "feature" // Old tracking parameter name, and may be obsolete.
    );

    /**
     * Injection point.
     */
    public static String sanitize(String url) {
        if (SharedYouTubeSettings.SANITIZE_SHARED_LINKS.get()) {
            url = sanitizer.sanitizeUrlString(url);
        }

        if (SharedYouTubeSettings.REPLACE_MUSIC_LINKS_WITH_YOUTUBE.get()) {
            url = url.replace("music.youtube.com", "youtube.com");
        }

        return url;
    }
}

File: patches/src/main/kotlin/app/morphe/patches/youtube/misc/privacy/SanitizeSharingLinksPatch.kt
package app.morphe.patches.youtube.misc.privacy

import app.morphe.patches.shared.misc.privacy.sanitizeSharingLinksPatch
import app.morphe.patches.youtube.misc.extension.sharedExtensionPatch
import app.morphe.patches.youtube.misc.settings.PreferenceScreen
import app.morphe.patches.youtube.misc.settings.settingsPatch

@Suppress("unused")
val sanitizeSharingLinksPatch = sanitizeSharingLinksPatch(
    block = {
        dependsOn(
            sharedExtensionPatch,
            settingsPatch,
        )

        compatibleWith(
            "com.google.android.youtube"(
                "20.14.43",
                "20.21.37",
                "20.31.42",
                "20.46.41",
            )
        )
    },
    preferenceScreen = PreferenceScreen.MISC
)
